### PR TITLE
Add ConVar changed callback

### PR DESCRIPTION
**Pull request by F1F7Y**
Allows modders to add a convar changed callback in their `mod.json`

ConVar def example:
```json
{
    "Name": "CALLBACK_TEST",
    "DefaultValue": "0",
    "Function": "TestConVar_Callback",
    "Context": "CLIENT"
}
```

Script:
```cpp
global function TestConVar_Callback

void function TestConVar_Callback( string oldValue, string newValue ) {
    printt("OldValue of CALLBACKS_TEST:", oldValue )
    printt("NewValue of CALLBACKS_TEST:", newValue )
}
```

### CODE REVIEW

**Comment by F1F7Y** [comment]
NOTE: There's a lot of fuckery in this which should be fixed before merge

**Comment by ASpoonPlaysGames** [comment]
Personally I would prefer if the JSON was structured more similarly to script callbacks, for example:

```json
{
    "Name": "CALLBACK_TEST",
    "DefaultValue": "0",
    "ClientCallback": "TestConVar_Callback",
    "UICallback": "UI_TestConVar_Callback",
    "ServerCallback": "SV_TestConVar_Callback"
}
```
This allows different callbacks to be defined per VM, as well as being more understandable than `"Function"`

**Comment by ASpoonPlaysGames** [comment]
I do like the idea though, callbacks for convar changes would be very nice

**Comment by F1F7Y** [comment]
> This allows different callbacks to be defined per VM, as well as being more understandable than `"Function"`

Yeah, this is so much better than what I have. Will implement later today or tomorrow

**Comment by ASpoonPlaysGames** [comment]
> definitely good, `<vm>ChangeCallback` would show the purpose best i think, e.g. `ClientChangeCallback` rather than just `ClientCallback`

yeah I agree, good call

File: NorthstarDLL/mods/modmanager.cpp [hunk 1]
```diff
@@ -460,7 +532,16 @@ void ModManager::LoadMods()
 			// behaviour is for defining same convar multiple times
 			if (!R2::g_pCVar->FindVar(convar->Name.c_str()))
 			{
-				new ConVar(convar->Name.c_str(), convar->DefaultValue.c_str(), convar->Flags, convar->HelpString.c_str());
+				new ConVar(
```
Comment (catornot) [inline NorthstarDLL/mods/modmanager.cpp:535]: maybe add a check if the convar actually has a callback defined?

File: NorthstarDLL/mods/modmanager.cpp [hunk 2]
```diff
@@ -352,6 +373,57 @@ auto ModConCommandCallback(const CCommand& command)
...
+			break;
+		}
+	}
+	if (!found)
+		return;
+
+	if (found->Callback.Function == "")
+		return;
+
+	spdlog::info(
```
Comment (catornot) [inline NorthstarDLL/mods/modmanager.cpp:410]: this could print too many logs if the convar would be edited often

**Review by catornot — COMMENTED** [review]
I tested the callbacks they mostly work fine except that convars with `NEVER_AS_STRING` flag aren't working.
logs:
```bash

] CALLBACK_TEST_NEVER_STRING tes
] CALLBACK_TEST e
[NORTHSTAR] [info] Running convar changed callback UI_TestConVar_Callback in UI for: CALLBACK_TEST
[SCRIPT UI] [info] oldValue of convar: test
[SCRIPT UI] [info] newValue of convar: e
] CALLBACK_TEST_NEVER_STRING 1
] CALLBACK_TEST_NEVER_STRING 10
] CALLBACK_TEST_NEVER_STRING 8
] CALLBACK_TEST 10
[NORTHSTAR] [info] Running convar changed callback UI_TestConVar_Callback in UI for: CALLBACK_TEST
[SCRIPT UI] [info] oldValue of convar: e
[SCRIPT UI] [info] newValue of convar: 10
] CALLBACK_TEST i
[NORTHSTAR] [info] Running convar changed callback UI_TestConVar_Callback in UI for: CALLBACK_TEST
[SCRIPT UI] [info] oldValue of convar: 10
[SCRIPT UI] [info] newValue of convar: i


```
mod that I made to test this
[cat_or_not.ConvarChangeCallbackTestMod.zip](https://github.com/R2Northstar/NorthstarLauncher/files/11023752/cat_or_not.ConvarChangeCallbackTestMod.zip)

another note is that I the current system is too limiting since you can only listen for changes on one vm.

**Comment by F1F7Y** [comment]
Use:
```json
{
	"Name": "CALLBACK_TEST",
	"DefaultValue": "0",
	"UIChangeCallback": "TestUI_ConVarCallback",
	"ClientChangeCallback": "TestClient_ConVarCallback",
	"ServerChangeCallback": "TestServer_ConVarCallback"
}
```

File: NorthstarDLL/mods/modmanager.cpp (hunk 1 above)
Comment (F1F7Y) [reply NorthstarDLL/mods/modmanager.cpp:526]: Is that needed? Imo there's no real benefit and it just clutters up the code

**Comment by F1F7Y** [comment]
Closing as i want to rework this ( soon :tm: )

**Pull request closed without merging**